Add support for nested URLRouter instances

matthiask opened the pull request:
Refs #870

Instead of modifying `scope["path"]` I introduced a new scope variable `"path_remaining"` for inner URLRouter instances, which -- if present -- is preferred by URLRouter to `"path"`.

The `test_url_router_nesting` test is modified to not expect positional AND keyword arguments, because Django also ignores all positional arguments as soon as there are any keyword arguments in URLconfs.

matthiask commented:
I added additional tests. Unfortunately, `path` cannot be used for the outer URLRouter easily because the ASGI app is callable, and therefore Django thinks that it is a view and appends a `$` to the regex.

andrewgodwin commented:
Ooh, that `path` thing is annoying, I forgot that was how it distinguished between views and includes. Ideally I'd like to make it work, though I accept that might be messy.

Is there any way you see of making it work? Or would it be too hard to reach inside and undo the `is_endpoint=True` damage to the regex?

matthiask commented:
[Here](https://github.com/matthiask/channels/tree/nested-urlrouter-path)'s a branch with an additional commit on top which makes the tests pass using `path()` in the outer URLRouter:

https://github.com/matthiask/channels/commit/cdd335be2e7c5a1d1049cc5b6f2b8a976cff07e9#diff-1a24e8059b03907de2c85bc441155308R85

It feels like a really bad idea though. The outer URLRouter would have to know that somewhere inside there's another ASGI app which wants to do some routing with the remaining path. The code in the commit above works only when 1. the inner ASGI app is an URLRouter and 2. there is no intermediate turtle in the stack of turtles. Introducing a way so that outer turtles know about inner turtles sounds very unasgilike (if that was a word).

Here's an alternative idea: Adding a `path()` function to `channels/routing.py` with support for overriding `is_endpoint` (and `re_path()` for consistency) -- no docs, and does not work with Django 1.11 yet:

https://github.com/matthiask/channels/commit/6d9830c9ec9e663da8bc38b45e715e1fcf3f882d

This also feels a bit bad. The `_path` function uses too many Django internals for my taste and I fear breakage down the road. It might be better to contribute this change to Django, but even almost 10 years after the initial contribution I find the contributing process a bit daunting and also we'd have to wait a long time since 1.11 is an LTS release.

andrewgodwin commented:
Yeah, both of those are too fragile for me, really - thanks for the investigative work though.

I'm just particularly uneasy about shipping this with `path` being a silent mysterious failure (obviously things are worse now, so we need this in some form, but I want to give it a decent level of thought).

What if, instead of trying to fix it, we just detect when it happens and raise an error? I suspect having an inner URLRouter detect that it is somewhere inside another URLRouter is quite easy (because of the presence of things in the `scope`), which means we just need to make sure that we can detect when there's no URL remaining to match on for the inner one and raise an error. This would also not work if someone deliberately made an inner URLRouter that just matched on `""` but... they should just not do that in that case.

matthiask commented:
Hmm. I'm not sure if I follow.

`URLRouter.__init__` does not have access to the `scope`, so we cannot signal to inner `URLRouter`s that they should check the outer routers' routes.

`URLRouter.__call__` of the inner router is never run when the outer `URLRouter` uses `path` because of the `$` in the regex, which seems only possible with fragile hacks.

I don't see a way around either changing Django's `path()` function or documenting this as a gotcha. It's a pity. `re_path()` isn't even strictly more powerful because of the converters that `path()` supports.

--- 

Here's a third version. This one modifies the routes passed to `URLRouter` upon initialization. Probably not better, since it uses the same trick as version 1 above, but only runs once.

https://github.com/matthiask/channels/commit/84c5dea7b365cc3733a099fa082758d1c854693f

andrewgodwin commented:
Oh, right, the inner one is never even called as it won't match. Hrmph.

That third version is not bad, it's only 3 lines of logic and it's not super-fragile. I would be alright with that.

matthiask commented:
I don't think it's a solution, because the third solution converts *all* routes to open-ended routes -- it apples `is_endpoint=False` everywhere, and does not look whether some inner ASGI app wants to do routing or not.

I'm still thinking about this issue, but haven't found a good solution yet. It's always too early (information isn't here yet) or too late (e.g. signaling to outer URLRouters that no inner ASGI app did any routing via `self.scope` -- but at this time, the event has already been processed in a possibly (probably) non-idempotent way)

andrewgodwin commented:
Let's not rush and try to make it as safe as possible :) People won't read documentation, of that I can assure you.

I'm much happier with a slightly hackish way of detecting errors, too, rather than a hackish fix. That said, I don't have any better ideas right now either...

matthiask commented:
Here's another idea: Add an attribute to ASGI apps, `does_path_routing`, which if `True` signals to outer URLRouter instances that it should not treat the app as an endpoint.

The change in version 3 above made the `test_url_router_path` fail because the first `path("")` did match everything. The new idea does not have this problem. 

Change added to this PR, but not documented yet. Relevant commit here: https://github.com/django/channels/pull/946/commits/04f82e95e1db897adee74ebd670382c0151f559f

matthiask commented:
Maybe `app.extensions = {"path_routing"}` or something would be better. (Similar to the channel layer extensions attribute.)

andrewgodwin commented:
That has the same issue as the earlier attempt to see if it was a URLRouter, though - it won't work if you've got middleware or similar in the way.

I think this is still probably the best approach, but I'm not sure this is the right place to jump to an `extensions` concept - maybe just have an underscored attribute instead?

matthiask commented:
Yes, it's true that middleware between URLRouter requires additional work, but setting an attribute seems less work (and also more correct) than inheriting URLRouter just for the `isinstance` check. I added a middleware without additional functionality to the [test](https://github.com/django/channels/pull/946/commits/948f855d729aabde96f17d50e1f8f33ec6f57962#diff-06c0af743b62c3e720fe92d0ce2b7f7bR129) for this reason.

Also, converted `extensions` back to `_path_routing`, and amended the test to `is True` instead of checking for truish values so that MagicMock works.

andrewgodwin commented:
Lovely, I think this will be the best we can get for now. Do you mind also adding a note to the routing documentation that mentions url nesting won't work properly with `path()` and urlrouters-inside-middleware?

matthiask commented:
I also reworded the kwargs/args discussion a bit (kwargs are mentioned first because they are preferred, and also mention that args and kwargs can't be mixed)

andrewgodwin commented:
Great - merged! Thanks for your work on this; with this and a couple other big things landed in the last week I'll get round to doing a new 2.1 release soon.

hoefling commented:
One question - is the copying of the scope necessary here?

https://github.com/django/channels/blob/984e9dc9cece66f7520bc9c1fd0583e9c60b1022/channels/routing.py#L113

I have a custom middleware that modifies the scope and the scope changes are not propagated to the consumers anymore. Example:

```python
class MyMiddleware:

    def __init__(self, scope: Scope, middleware: Middleware) -> None:
        self.scope = scope
        self.middleware = middleware
        self.inner = self.middleware.inner(scope)
        self.receive_orig = None
        self.send_orig = None

    def __call__(self, receive: Receiver, send: Sender) -> Middleware:
        self.receive_orig = receive
        return self.inner(self.receive, send)

    async def receive(self):
        """Overridden receive that clears stored credentials once the token expires."""
        if self.scope['token'] is not valid:
            self.scope.pop('token', None)
        return await self.receive_orig()

class MyConsumer(AsyncWebsocketConsumer):
    async def receive(self, text_data=None, bytes_data=None):
        if 'token' not in self.scope:
            await self.close()
        else:
            await self.send(text_data='ping')

Removing `token` has no effect in `MyConsumer`. Not sure if this also affects `SessionMiddleware`, but possible.

matthiask commented:
I did give this particular behavior a lot of thought, but probably not enough. `scope` isn't copied anywhere else, so if possible it shouldn't be copied here either.

I thought that path_remaining should not be visible to other routers and middleware, and the easiest way to ensure this was to make a shallow copy of the scope. 

The current tests all pass without the shallow copy, but I'd like to add some additional tests to annoy the code some more. I should be able to submit another pull request until tomorrow evening (UTC)

hoefling commented:
@matthiask let's wait for what @andrewgodwin has to say :-) First you clearly had your reasons to implement things as they are, second I'm not even sure the middleware should be allowed to modify the scope - it's just something I assumed based on Django's middleware hooking patterns. If the scope shouldn't be shared on the way in the whole middleware chain, it's also fine, I'll just move the middleware's logic to the consumer itself, maybe introduce some consumer mixin for the reusable code.

matthiask commented:
It's almost too easy to construct a failing test with the current code:

```
# @pytest.mark.xfail
def test_path_remaining():
    inner_router = URLRouter([
        url(r'^no-match/$', MagicMock(return_value=1)),
    ])
    test_app = MagicMock(return_value=2)
    outer_router = URLRouter([
        url(r'^prefix/', inner_router),
        url(r'^prefix/stuff/$', test_app),
    ])

    assert outer_router({"type": "http", "path": "/prefix/stuff/"}) == 2
```

The two changes that are necessary for this test to pass are the following

- Restore `path_remaining` after calling `route.callback(scope)` 
- Raise a `Resolver404` instead of a `ValueError` if no route of the current router matches, since this exception is handled by an outer URLRouter.

Unfortunately the `Resolver404` exception is not known to Daphne, otherwise this might also help with solving https://github.com/django/daphne/issues/165 (Return 404 for invalid routes)

The WIP code is available here: https://github.com/matthiask/channels/commit/bd9bda78a8ad876493357b923012ec9feb60f6a3

andrewgodwin commented:
I had always in my head assumed that scope was something that was immutable once it was handed down through the ASGI first-call, so I was for adding `copy` in though. You are right, however, that we should at least be consistent about it - either make all other middlewares we ship also copy, or remove the copy here.

The reason I would probably vote for everything copying is that modifying scope in a middleware like in the example above feels wrong - it's not meant to be something you check dynamically, but more like a Django HttpRequest where you can tack on attributes but you're not meant to change it in a major way. Thoughts?

matthiask commented:
I think copying makes more sense, yes. The `scope` comes from outside; if a middleware or a router wants to save additional data somewhere, it can just add attributes to `self`.

I force updated the branch referenced above; to me the code looks much better like this:
https://github.com/django/channels/compare/master...matthiask:more-urlrouter-nesting

Should I prepare a pull request implementing the scope copying change? It would involve changing a few places in the documentation too:

```
~/Projects/channels/channels$ git grep -E 'scope\[.* = '
channels/auth.py:    scope["user"] = user
channels/auth.py:        scope["user"] = AnonymousUser()
channels/auth.py:            scope["user"] = SimpleLazyObject(lambda: async_to_sync(get_user)(scope))
channels/http.py:            self.scope["headers"] = [
channels/sessions.py:                scope["cookies"] = parse_cookie(value.decode("ascii"))
channels/sessions.py:            scope["cookies"] = {}
channels/sessions.py:        self.scope["session"] = self.middleware.session_store(session_key)
channels/staticfiles.py:            scope["static_base_url"] = self.base_url
docs/topics/authentication.rst:            scope["user"] = User.objects.get(id=int(scope["query_string"]))
docs/topics/sessions.rst:            self.scope["session"]["seed"] = random.randint(1, 1000)
~/Projects/channels/channels$ 
```

(Idle thought: In some ways `scope` is a bit misleading to me, because it sounds like it would be the preferred place to put variables. It is more similar to an `environ`ment for inner ASGI apps. Modifying a scope is what we do all the time when programming, modifying the environment is a thing which requires more thought.)

hoefling commented:
Nice to see a decision here! I would also request some explaining note regarding it in the docs as it seems I'm not the only one being misled, see the PR review [here](https://github.com/django/channels/pull/973/files#r174378332).

ovekaaven commented:
This seems to introduce a regression for me:

```
2018-04-01 00:44:45,441 - ERROR - ws_protocol - Traceback (most recent call last):
  File "/home/ovek/django-xmpp-server/example/env/lib/python3.6/site-packages/daphne/ws_protocol.py", line 76, in onConnect
    "subprotocols": subprotocols,
  File "/home/ovek/django-xmpp-server/example/env/lib/python3.6/site-packages/daphne/server.py", line 184, in create_application
    application_instance = self.application(scope=scope)
  File "/home/ovek/django-xmpp-server/example/chatbox/channels/staticfiles.py", line 41, in __call__
    return self.application(scope)
  File "/home/ovek/django-xmpp-server/example/chatbox/channels/routing.py", line 51, in __call__
    return self.application_mapping[scope["type"]](scope)
  File "/home/ovek/django-xmpp-server/example/chatbox/channels/routing.py", line 119, in __call__
    "kwargs": {**outer.get("kwargs", {}), **kwargs},
TypeError: 'NoneType' object is not callable
```

I'm using includes in urls:

```
ws_urls = [
    url(r'^bind/$', WSConsumer, name='ws'),
]

application = ProtocolTypeRouter({
    'http': URLRouter([
        url(r'', AsgiHandler)
    ]),
    'websocket': URLRouter([
        url(r'^chat/', include(ws_urls))
    ])
})

```

matthiask commented:
@ovekaaven Indeed it does. There is no test in channels for `include()` inside `URLRouter()`, so this has not been detected earlier. 

Failing test here: https://github.com/matthiask/channels/commit/accf656a717e606f4043da26181b9a2ff1acf565  I'll see whether I can fix this.

andrewgodwin commented:
I did not intend for includes to work, which is why there is no test for them.

ovekaaven commented:
If you don't want them to work, then you should at least update the documentation.

andrewgodwin commented:
I'm not saying I don't want them to, just that I never got round to documenting and testing either way. It looks like support may be possible given #996, it's now a question of if it makes sense to offer two ways of nesting.
